packages/scripts/src/admin-proposals/funding/0_Propose.ts

```diff
@@ -0,0 +1,141 @@
+    throw new Error("PROPOSAL_URL must be a valid URL");
+  }
+
+  return {
+    proposerSigner: await getProposerSigner(env),
+    recipient: env.RECIPIENT,
+    tokenAddress: env.TOKEN || (await getAddress("VotingToken", 1)), // Default to UMA if token not provided.
+    amount: Number(env.AMOUNT),
+    proposalUrl: env.PROPOSAL_URL,
+    trace: env.TRACE === "1",
+  };
+}
+
+async function getAndVerifyFundingAmount(
+  params: ProposalParams
+): Promise<{ token: ERC20Ethers; fundingAmount: BigNumber }> {
+  // Get amount scaled by decimals.
+  const token = await getContractInstance<ERC20Ethers>("ERC20", params.tokenAddress);
+  const decimals = await token.decimals();
+  const fundingAmount = BigNumber.from(params.amount).mul(BigNumber.from(10).pow(decimals));
+
+  // Verify that GovernorV2 has sufficient balance.
+  const governorV2Address = await getAddress("GovernorV2", 1);
+  const governorBalance = await token.balanceOf(governorV2Address);
+  if (governorBalance.lt(fundingAmount)) throw new Error("Insufficient balance to fund proposal");
+
+  return { token, fundingAmount };
+}
+
+async function verifyProposerBond(params: ProposalParams, proposerV2: ProposerV2Ethers): Promise<void> {
+  // Approve required bond amount if necessary.
+  const votingToken = await getContractInstance<VotingTokenEthers>("VotingToken");
+  const bond = await proposerV2.bond();
+  const allowance = await votingToken.allowance(proposerAddress, proposerV2.address);
+  if (allowance.lt(bond)) {
+    console.log("Approving proposer bond");
+    const approveTx = await votingToken.connect(params.proposerSigner).approve(proposerV2.address, bond);
+    await approveTx.wait();
+  }
+
+  // Check proposer balance.
+  const balance = await votingToken.balanceOf(proposerAddress);
+  if (balance.lt(bond)) throw new Error("Insufficient proposer balance");
+}
+
+async function main() {
+  const params = await initProposalParams(process.env);
+
+  // Get raw funding amount and verify the governor has sufficient balance.
+  const { token, fundingAmount } = await getAndVerifyFundingAmount(params);
+
+  // Make sure proposer has sufficient bond balance and allowance.
+  const proposerV2 = await getContractInstance<ProposerV2Ethers>("ProposerV2");
+  await verifyProposerBond(params, proposerV2);
+
+  // Construct and propose funding transaction.
+  const adminProposalTransactions: AdminProposalTransaction[] = [
+    {
+      to: params.tokenAddress,
+      value: "0",
+      data: token.interface.encodeFunctionData("transfer", [params.recipient, fundingAmount]),
+    },
+  ];
+  console.log("Sending proposal transactions to the proposer");
+  const ancillaryData = `title: Admin funding proposal, proposalUrl: "${params.proposalUrl}"`;
+  const txn = await proposerV2
+    .connect(params.proposerSigner)
+    .propose(adminProposalTransactions, ethers.utils.toUtf8Bytes(ancillaryData));
+  await txn.wait();
+  if (params.trace) await hre.run("trace", { hash: txn.hash });
```

Comment:
What is this? I've never heard of HRE trace? Does this just print out a trace or something?

Reply:
yes, it does the same as running `hardhat trace` on command line. I find this helpful to check internal calls and emitted logs